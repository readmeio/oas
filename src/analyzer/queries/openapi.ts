import type { OASDocument } from '../../rmoas.types';

import Oas from '../..';
import { query, refizePointer } from '../util';

/**
 * Determine if a given API definition uses the `additionalProperties` schema property.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#schema-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object}
 */
export function additionalProperties(definition: OASDocument) {
  return query(['$..additionalProperties'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine if a given API definition utilizes `callbacks`.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callbackObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callbackObject}
 */
export function callbacks(definition: OASDocument) {
  return query(['$.components.callbacks', '$.paths..callbacks'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine if a given API definition has circular refs.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#schema-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object}
 */
export async function circularRefs(definition: OASDocument) {
  // Dereferencing will update the passed in variable, which we don't want to do, so we
  // instantiated `Oas` with a clone.
  const oas = new Oas(JSON.parse(JSON.stringify(definition)));
  await oas.dereference();

  const results = oas.getCircularReferences();

  results.sort();
  return results;
}

/**
 * Determine if a given API definition utilizes discriminators.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#discriminatorObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#discriminatorObject}
 */
export function discriminators(definition: OASDocument) {
  return query(['$..discriminator'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine if a given API definition utilizes `links`.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#linkObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#linkObject}
 */
export function links(definition: OASDocument) {
  return query(['$..links'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine all of the available media types used within an API definition.
 *
 * @todo This query currently picks up false positives if there is an object named `content`.
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#request-body-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#request-body-object}
 */
export function mediaTypes(definition: OASDocument) {
  const results = Array.from(
    new Set(
      query(['$..paths..content'], definition)
        .map(res => {
          // This'll transform `results`, which looks like `[['application/json'], ['text/xml']]`
          // into `['application/json', 'text/xml']`.
          return Object.keys(res.value);
        })
        .flat()
    )
  );

  results.sort();
  return results;
}

/**
 * Determine if a given API definition uses parameter serialization.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#parameter-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#parameter-object}
 */
export function parameterSerialization(definition: OASDocument) {
  return query(['$..parameters[*].style^'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine if a given API definition utilizes schema polymorphism and/of interitance.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#schema-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object}
 */
export function polymorphism(definition: OASDocument) {
  const results = Array.from(
    new Set(query(['$..allOf^', '$..anyOf^', '$..oneOf^'], definition).map(res => refizePointer(res.pointer)))
  );

  results.sort();
  return results;
}

/**
 * Determine every kind of security type that a given API definition has documented.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#securitySchemeObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#securitySchemeObject}
 */
export function securityTypes(definition: OASDocument) {
  return Array.from(new Set(query(['$.components.securitySchemes..type'], definition).map(res => res.value as string)));
}

/**
 * Determine if a given API definition utilizes server variables.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#serverVariableObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#serverVariableObject}
 */
export function serverVariables(definition: OASDocument) {
  return query(['$.servers..variables^'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine how many operations are defined in a given API definition.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#operationObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#operationObject}
 */
export function totalOperations(definition: OASDocument) {
  return query(['$..paths[*]'], definition)
    .map(res => Object.keys(res.value))
    .flat().length;
}

/**
 * Determine if a given API definition utilizes `webhooks` support in OpenAPI 3.1.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#oasObject}
 */
export function webhooks(definition: OASDocument) {
  return query(['$.webhooks[*]'], definition).map(res => refizePointer(res.pointer));
}

/**
 * Determine if a given API definition has XML schemas, payloads, or responses.
 *
 * @todo detect `+xml` media types
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#xmlObject}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#xmlObject}
 */
export function xml(definition: OASDocument) {
  return query(
    [
      '$.components.schemas..xml^',
      '$..parameters..xml^',
      '$..requestBody..xml^',

      "$..requestBody..['application/xml']",
      "$..requestBody..['application/xml-external-parsed-entity']",
      "$..requestBody..['application/xml-dtd']",
      "$..requestBody..['text/xml']",
      "$..requestBody..['text/xml-external-parsed-entity']",
      '$..requestBody.content[?(@property.match(/\\+xml$/i))]',

      "$..responses..['application/xml']",
      "$..responses..['application/xml-external-parsed-entity']",
      "$..responses..['application/xml-dtd']",
      "$..responses..['text/xml']",
      "$..responses..['text/xml-external-parsed-entity']",
      '$..responses[*].content[?(@property.match(/\\+xml$/i))]',
    ],
    definition
  ).map(res => refizePointer(res.pointer));
}
