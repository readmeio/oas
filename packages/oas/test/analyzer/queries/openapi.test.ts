import type { OASDocument } from '../../../src/types.js';

import callbacks from '@readme/oas-examples/3.0/json/callbacks.json' with { type: 'json' };
import complexNesting from '@readme/oas-examples/3.0/json/complex-nesting.json' with { type: 'json' };
import discriminators from '@readme/oas-examples/3.0/json/discriminators.json' with { type: 'json' };
import links from '@readme/oas-examples/3.0/json/link-example.json' with { type: 'json' };
import commonParameters from '@readme/oas-examples/3.0/json/parameters-common.json' with { type: 'json' };
import petstore from '@readme/oas-examples/3.0/json/petstore.json' with { type: 'json' };
import readmeLegacy from '@readme/oas-examples/3.0/json/readme-legacy.json' with { type: 'json' };
import additionalProperties from '@readme/oas-examples/3.0/json/schema-additional-properties.json' with { type: 'json' };
import circular from '@readme/oas-examples/3.0/json/schema-circular.json' with { type: 'json' };
import security from '@readme/oas-examples/3.0/json/security.json' with { type: 'json' };
import serverVariables from '@readme/oas-examples/3.0/json/server-variables.json' with { type: 'json' };
import parameterStyles from '@readme/oas-examples/3.1/json/parameters-style.json' with { type: 'json' };
import readme from '@readme/oas-examples/3.1/json/readme.json' with { type: 'json' };
import trainTravel from '@readme/oas-examples/3.1/json/train-travel.json' with { type: 'json' };
import webhooks from '@readme/oas-examples/3.1/json/webhooks.json' with { type: 'json' };
import { describe, expect, it } from 'vitest';

import * as QUERIES from '../../../src/analyzer/queries/openapi.js';
import responses from '../../__datasets__/responses.json' with { type: 'json' };

describe('analyzer queries (OpenAPI)', () => {
  describe('additionalProperties', () => {
    it('should discover `additionalProperties` usage within a definition that has it', () => {
      expect(QUERIES.additionalProperties(additionalProperties as OASDocument)).toStrictEqual([
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: $ref, simple`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: $ref, with $ref`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: false` and no other properties/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: true`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: { type: array, items: { type: integer } }`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: { type: integer }`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: { type: object, properties: ... }` and custom title/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalProperties: { type: object, properties: ... }`/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalPropeties` within an allOf/allOf/0/additionalProperties',
        '#/paths/~1post/post/requestBody/content/application~1json/schema/properties/object with `additionalPropeties` within an allOf/allOf/1/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: $ref, simple`/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: $ref, with $ref`/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: true`/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: { type: array, items: { type: integer } }`/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: { type: integer }`/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: { type: object, properties: ... }` and custom title/additionalProperties',
        '#/paths/~1post/post/responses/200/content/application~1json/schema/properties/object with `additionalProperties: { type: object, properties: ... }`/additionalProperties',
      ]);
    });

    it('should discover `additionalProperties` usage within `$ref` pointers', () => {
      expect(QUERIES.additionalProperties(complexNesting as OASDocument)).toStrictEqual([
        '#/components/schemas/ObjectOfAdditionalPropertiesObjectPolymorphism/additionalProperties',
      ]);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.additionalProperties(security as OASDocument)).toHaveLength(0);
    });
  });

  describe('callbacks', () => {
    it('should discover `callbacks` usage within a definition that has it', () => {
      expect(QUERIES.callbacks(callbacks as OASDocument)).toStrictEqual(['#/paths/~1streams/post/callbacks']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.callbacks(readmeLegacy as OASDocument)).toHaveLength(0);
    });

    it('should not flag schema properties named `callbacks` as a callback', () => {
      expect(
        readme.paths['/branches/{branch}/reference'].post.requestBody.content['application/json'].schema.properties.api
          .properties.stats.properties,
      ).toHaveProperty('callbacks');

      expect(QUERIES.callbacks(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('circularRefs', () => {
    it('should determine if a definition has circular refs when it does', async () => {
      await expect(QUERIES.circularRefs(circular as OASDocument)).resolves.toStrictEqual([
        '#/components/schemas/MultiPart/properties/parent',
        '#/components/schemas/ZoneOffset/properties/rules',
      ]);
    });

    it("should not find where it doesn't exist", async () => {
      await expect(QUERIES.circularRefs(readme as OASDocument)).resolves.toHaveLength(0);
    });
  });

  describe('commonParameters', () => {
    it('should discover common parameters usage within a definition that has it', () => {
      expect(QUERIES.commonParameters(commonParameters as OASDocument)).toStrictEqual([
        '#/paths/~1anything~1{id}/parameters',
        '#/paths/~1anything~1{id}~1override/parameters',
        '#/paths/~1anything~1{id}~1{action}/parameters',
        '#/paths/~1anything~1{id}~1{action}~1{id}/parameters',
      ]);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.commonParameters(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('discriminators', () => {
    it('should discover `discriminator` usage within a definition that has it', () => {
      expect(QUERIES.discriminators(discriminators as OASDocument)).toStrictEqual([
        '#/components/schemas/BaseVehicle/discriminator',
        '#/components/schemas/Pet/discriminator',
        '#/paths/~1discriminator-with-mapping/patch/requestBody/content/application~1json/schema/discriminator',
        '#/paths/~1discriminator-with-no-mapping/patch/requestBody/content/application~1json/schema/discriminator',
        '#/paths/~1improper-discriminator-placement/patch/requestBody/content/application~1json/schema/properties/connector_properties/discriminator',
        '#/paths/~1mapping-of-schema-names/patch/requestBody/content/application~1json/schema/discriminator',
        '#/paths/~1mapping-with-duplicate-schemas/patch/requestBody/content/application~1json/schema/discriminator',
        '#/paths/~1oneof-allof-top-level-disc/patch/requestBody/content/application~1json/schema/discriminator',
        '#/paths/~1potentially-undefined-formData/post/requestBody/content/application~1json/schema/discriminator',
      ]);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.discriminators(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('fileSize', () => {
    it('should calculate the size of the definition in its raw form', async () => {
      await expect(QUERIES.fileSize(trainTravel as OASDocument)).resolves.toStrictEqual({
        raw: 0.03,
        dereferenced: 0.15,
      });
    });
  });

  describe('links', () => {
    it('should discover `links` usage within a definition that has it', () => {
      expect(QUERIES.links(links as OASDocument)).toStrictEqual([
        '#/components/links',
        '#/paths/~12.0~1repositories~1{username}/get/responses/200/links',
        '#/paths/~12.0~1repositories~1{username}~1{slug}/get/responses/200/links',
        '#/paths/~12.0~1repositories~1{username}~1{slug}~1pullrequests~1{pid}/get/responses/200/links',
        '#/paths/~12.0~1users~1{username}/get/responses/200/links',
      ]);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.links(readmeLegacy as OASDocument)).toHaveLength(0);
    });

    it('should not flag schema properties named `links` as a link', () => {
      expect(
        readme.paths['/images'].post.responses[201].content['application/json'].schema.properties.data.properties,
      ).toHaveProperty('links');

      expect(QUERIES.links(readme as OASDocument)).toStrictEqual([]);
    });
  });

  describe('mediaTypes', () => {
    it('should tally all of the media types used within a definition', () => {
      expect(QUERIES.mediaTypes(petstore as OASDocument)).toStrictEqual([
        'application/json',
        'application/x-www-form-urlencoded',
        'application/xml',
        'multipart/form-data',
      ]);
    });
  });

  describe('parameter serialization', () => {
    it('should discover parameter serialization usage within a definition that has it', () => {
      expect(QUERIES.parameterSerialization(parameterStyles as unknown as OASDocument)).toStrictEqual([
        '#/paths/~1anything~1headers~1simple/get/parameters/0',
        '#/paths/~1anything~1headers~1simple/get/parameters/1',
        '#/paths/~1anything~1headers~1simple/get/parameters/2',
        '#/paths/~1anything~1headers~1simple/post/parameters/0',
        '#/paths/~1anything~1headers~1simple/post/parameters/1',
        '#/paths/~1anything~1headers~1simple/post/parameters/2',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/get/parameters/0',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/get/parameters/1',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/get/parameters/2',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/post/parameters/0',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/post/parameters/1',
        '#/paths/~1anything~1path~1label~1{primitive}~1{array}~1{object}/post/parameters/2',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/get/parameters/0',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/get/parameters/1',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/get/parameters/2',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/post/parameters/0',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/post/parameters/1',
        '#/paths/~1anything~1path~1matrix~1{primitive}~1{array}~1{object}/post/parameters/2',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/get/parameters/0',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/get/parameters/1',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/get/parameters/2',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/post/parameters/0',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/post/parameters/1',
        '#/paths/~1anything~1path~1simple~1{primitive}~1{array}~1{object}/post/parameters/2',
        '#/paths/~1anything~1query~1deepObject/get/parameters/0',
        '#/paths/~1anything~1query~1form/get/parameters/0',
        '#/paths/~1anything~1query~1form/get/parameters/1',
        '#/paths/~1anything~1query~1form/get/parameters/2',
        '#/paths/~1anything~1query~1form/post/parameters/0',
        '#/paths/~1anything~1query~1form/post/parameters/1',
        '#/paths/~1anything~1query~1form/post/parameters/2',
        '#/paths/~1anything~1query~1pipeDelimited/get/parameters/0',
        '#/paths/~1anything~1query~1pipeDelimited/get/parameters/1',
        '#/paths/~1anything~1query~1spaceDelimited/get/parameters/0',
        '#/paths/~1anything~1query~1spaceDelimited/get/parameters/1',
        '#/paths/~1cookies#formExploded/get/parameters/0',
        '#/paths/~1cookies#formExploded/get/parameters/1',
        '#/paths/~1cookies#formExploded/get/parameters/2',
        '#/paths/~1cookies#formNonExploded/get/parameters/0',
        '#/paths/~1cookies#formNonExploded/get/parameters/1',
        '#/paths/~1cookies#formNonExploded/get/parameters/2',
      ]);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.parameterSerialization(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('polymorphism', () => {
    it('should determine if a definition uses schema polymorphism', () => {
      expect(QUERIES.polymorphism(readme as OASDocument)).toMatchSnapshot();
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.polymorphism(petstore as OASDocument)).toHaveLength(0);
    });
  });

  describe('securityTypes', () => {
    it('should tally all of the security types used within a definition', () => {
      expect(QUERIES.securityTypes(security as OASDocument)).toStrictEqual([
        'apiKey',
        'http',
        'oauth2',
        'openIdConnect',
      ]);
    });
  });

  describe('server variables', () => {
    it('should determine if a definition uses server variables when it does', () => {
      expect(QUERIES.serverVariables(serverVariables)).toStrictEqual(['#/servers/0', '#/servers/1', '#/servers/2']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.serverVariables(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('totalOperations', () => {
    it('should count the total operations used within a definition', () => {
      expect(QUERIES.totalOperations(readme as OASDocument)).toBe(54);
    });
  });

  describe('webhooks', () => {
    it('should determine if a definition uses webhooks when it does', () => {
      expect(QUERIES.webhooks(webhooks as OASDocument)).toStrictEqual(['#/webhooks/newPet']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.webhooks(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('xml', () => {
    it('should determine if a definition has XML', () => {
      expect(QUERIES.xml(petstore as OASDocument)).toStrictEqual([
        '#/components/schemas/Category',
        '#/components/schemas/Order',
        '#/components/schemas/Pet',
        '#/components/schemas/Pet/properties/photoUrls',
        '#/components/schemas/Pet/properties/tags',
        '#/components/schemas/Tag',
        '#/components/schemas/User',
        '#/paths/~1pet~1findByStatus/get/responses/200/content/application~1xml',
        '#/paths/~1pet~1findByTags/get/responses/200/content/application~1xml',
        '#/paths/~1pet~1{petId}/get/responses/200/content/application~1xml',
        '#/paths/~1store~1order/post/responses/200/content/application~1xml',
        '#/paths/~1store~1order~1{orderId}/get/responses/200/content/application~1xml',
        '#/paths/~1user~1login/get/responses/200/content/application~1xml',
        '#/paths/~1user~1{username}/get/responses/200/content/application~1xml',
      ]);
    });

    it('should discover `+xml` vendor suffixes', () => {
      expect(
        QUERIES.xml({
          paths: {
            '/anything': {
              get: {
                responses: {
                  '200': {
                    description: 'successful operation',
                    content: {
                      'text/plain+xml': {
                        schema: {
                          type: 'array',
                          items: {
                            type: 'string',
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        } as any),
      ).toStrictEqual(['#/paths/~1anything/get/responses/200/content/text~1plain+xml']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.xml(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('xmlSchemas', () => {
    it('should determine if a definition uses the XML object', () => {
      expect(QUERIES.xmlSchemas(trainTravel as OASDocument)).toStrictEqual([
        '#/components/schemas/Booking',
        '#/components/schemas/Problem',
        '#/components/schemas/Station',
        '#/components/schemas/Trip',
        '#/components/schemas/Wrapper-Collection',
      ]);
    });

    it('should not detect the `+xml` vendor suffix as an XML object', () => {
      expect(QUERIES.xmlSchemas(responses as unknown as OASDocument)).toHaveLength(0);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.xmlSchemas(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('xmlRequests', () => {
    it('should determine if a definition has XML payloads', () => {
      expect(QUERIES.xmlRequests(trainTravel as OASDocument)).toStrictEqual([
        '#/paths/~1bookings/post/requestBody/content/application~1xml',
      ]);
    });

    it('should discover `+xml` vendor suffixes', () => {
      expect(
        QUERIES.xmlRequests({
          paths: {
            '/anything': {
              post: {
                requestBody: {
                  required: true,
                  content: {
                    'text/plain+xml': {
                      schema: {
                        type: 'array',
                        items: {
                          type: 'string',
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        } as any),
      ).toStrictEqual(['#/paths/~1anything/post/requestBody/content/text~1plain+xml']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.xmlRequests(readme as OASDocument)).toHaveLength(0);
    });
  });

  describe('xmlResponses', () => {
    it('should determine if a definition has XML responses', () => {
      expect(QUERIES.xmlResponses(petstore as OASDocument)).toStrictEqual([
        '#/paths/~1pet~1findByStatus/get/responses/200/content/application~1xml',
        '#/paths/~1pet~1findByTags/get/responses/200/content/application~1xml',
        '#/paths/~1pet~1{petId}/get/responses/200/content/application~1xml',
        '#/paths/~1store~1order/post/responses/200/content/application~1xml',
        '#/paths/~1store~1order~1{orderId}/get/responses/200/content/application~1xml',
        '#/paths/~1user~1login/get/responses/200/content/application~1xml',
        '#/paths/~1user~1{username}/get/responses/200/content/application~1xml',
      ]);
    });

    it('should discover `+xml` vendor suffixes', () => {
      expect(
        QUERIES.xmlResponses({
          paths: {
            '/anything': {
              get: {
                responses: {
                  '200': {
                    description: 'successful operation',
                    content: {
                      'text/plain+xml': {
                        schema: {
                          type: 'array',
                          items: {
                            type: 'string',
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        } as any),
      ).toStrictEqual(['#/paths/~1anything/get/responses/200/content/text~1plain+xml']);
    });

    it("should not find where it doesn't exist", () => {
      expect(QUERIES.xmlResponses(readme as OASDocument)).toHaveLength(0);
    });
  });
});
