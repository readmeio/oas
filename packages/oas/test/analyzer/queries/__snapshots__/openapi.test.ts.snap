// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyzer queries (OpenAPI) > polymorphism > should determine if a definition uses schema polymorphism 1`] = `
[
  "#/paths/~1branches/get/responses/200/content/application~1json/schema",
  "#/paths/~1branches/post/requestBody/content/application~1json/schema",
  "#/paths/~1branches/post/responses/201/content/application~1json/schema",
  "#/paths/~1branches~1{branch}/get/responses/200/content/application~1json/schema",
  "#/paths/~1branches~1{branch}/patch/requestBody/content/application~1json/schema",
  "#/paths/~1branches~1{branch}/patch/responses/200/content/application~1json/schema",
  "#/paths/~1branches~1{branch}~1categories~1{section}~1{title}~1pages/get/responses/200/content/application~1json/schema/properties/data/items/properties/slug",
  "#/paths/~1branches~1{branch}~1guides/post/requestBody/content/application~1json/schema/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1guides/post/requestBody/content/application~1json/schema/properties/slug",
  "#/paths/~1branches~1{branch}~1guides/post/responses/201/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1guides/post/responses/201/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/get/responses/200/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/get/responses/200/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/patch/requestBody/content/application~1json/schema/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/patch/requestBody/content/application~1json/schema/properties/slug",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/patch/responses/200/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1guides~1{slug}/patch/responses/200/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1branches~1{branch}~1reference/post/requestBody/content/application~1json/schema/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1reference/post/requestBody/content/application~1json/schema/properties/slug",
  "#/paths/~1branches~1{branch}~1reference/post/responses/201/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1reference/post/responses/201/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/get/responses/200/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/get/responses/200/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/patch/requestBody/content/application~1json/schema/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/patch/requestBody/content/application~1json/schema/properties/slug",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/patch/responses/200/content/application~1json/schema/properties/data/properties/content/properties/next/properties/pages/items",
  "#/paths/~1branches~1{branch}~1reference~1{slug}/patch/responses/200/content/application~1json/schema/properties/data/properties/slug",
  "#/paths/~1changelogs~1{identifier}/delete/parameters/0/schema",
  "#/paths/~1changelogs~1{identifier}/get/parameters/0/schema",
  "#/paths/~1changelogs~1{identifier}/patch/parameters/0/schema",
  "#/paths/~1projects~1{subdomain}~1apikeys/post/requestBody/content/application~1json/schema/properties/label",
  "#/paths/~1projects~1{subdomain}~1apikeys~1{api_key_id}/patch/requestBody/content/application~1json/schema/properties/label",
]
`;
